src/Std/Data/Iterators/Lemmas/Consumers/Collect.lean

```diff
@@ -0,0 +1,111 @@
+/-
+Copyright (c) 2025 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul Reichert
+-/
+prelude
+import Std.Data.Iterators.Consumers.Access
+import Std.Data.Iterators.Consumers.Collect
+import Std.Data.Iterators.Lemmas.Basic
+import Std.Data.Iterators.Lemmas.Consumers.Monadic.Collect
+
+namespace Std.Iterators
+
+theorem Iter.toArray_eq_toArray_toIterM {α β} [Iterator α Id β] [Finite α Id] [IteratorCollect α Id]
+    [LawfulIteratorCollect α Id] {it : Iter (α := α) β} :
+    it.toArray = it.toIterM.toArray :=
+  rfl
+
+theorem Iter.toList_eq_toList_toIterM {α β} [Iterator α Id β] [Finite α Id] [IteratorCollect α Id]
+    [LawfulIteratorCollect α Id] {it : Iter (α := α) β} :
+    it.toList = it.toIterM.toList :=
+  rfl
+
+theorem Iter.toListRev_eq_toListRev_toIterM {α β} [Iterator α Id β] [Finite α Id]
+    {it : Iter (α := α) β} :
+    it.toListRev = it.toIterM.toListRev :=
+  rfl
+
+@[simp]
+theorem IterM.toList_toPureIter {α β} [Iterator α Id β] [Finite α Id] [IteratorCollect α Id]
+    {it : IterM (α := α) Id β} :
+    it.toPureIter.toList = it.toList :=
+  rfl
+
+@[simp]
+theorem IterM.toListRev_toPureIter {α β} [Iterator α Id β] [Finite α Id]
+    {it : IterM (α := α) Id β} :
+    it.toPureIter.toListRev = it.toListRev :=
+  rfl
+
+theorem Iter.toList_toArray {α β} [Iterator α Id β] [Finite α Id] [IteratorCollect α Id]
+    [LawfulIteratorCollect α Id] {it : Iter (α := α) β} :
+    it.toArray.toList = it.toList := by
+  simp only [toArray_eq_toArray_toIterM, toList_eq_toList_toIterM, ← IterM.toList_toArray,
+    Id.map_eq]
+
+theorem Iter.toArray_toList {α β} [Iterator α Id β] [Finite α Id] [IteratorCollect α Id]
+    [LawfulIteratorCollect α Id] {it : Iter (α := α) β} :
+    it.toList.toArray = it.toArray := by
+  simp only [toArray_eq_toArray_toIterM, toList_eq_toList_toIterM, ← IterM.toArray_toList,
+    Id.map_eq]
+
+theorem Iter.toListRev_eq {α β} [Iterator α Id β] [Finite α Id] [IteratorCollect α Id]
+    [LawfulIteratorCollect α Id] {it : Iter (α := α) β} :
+    it.toListRev = it.toList.reverse := by
+  simp [Iter.toListRev_eq_toListRev_toIterM, Iter.toList_eq_toList_toIterM, IterM.toListRev_eq]
+
+theorem Iter.toArray_of_step {α β} [Iterator α Id β] [Finite α Id] [IteratorCollect α Id]
```

Comment:
I'm not really sold on using `of` to describe something in the goal. Would `toArray_eq_match_step` work?

Reply:
I like it!